src/mon/MonCap.cc

```diff
@@ -601,14 +601,12 @@ struct MonCapParser : qi::grammar<Iterator, MonCap()>
 
 bool MonCap::parse(const string& str, ostream *err)
 {
-  string s = str;
-  string::iterator iter = s.begin();
-  string::iterator end = s.end();
-
-  MonCapParser<string::iterator> g;
-  bool r = qi::parse(iter, end, g, *this);
-  //MonCapGrant foo;
-  //bool r = qi::phrase_parse(iter, end, g, ascii::space, foo);
+  auto iter = str.begin();
+  auto end = str.end();
+
+  MonCapParser<string::const_iterator> pexp;
+  bool r = qi::phrase_parse(iter, end, pexp, ascii::space, *this);
```

Comment:
why `s/parse/phrase_parse/` here?

Reply:
Because `parse` doesn't accept const `iter` and const `end`. Only non-const. This was the reason, a copy of `str` was introduced in this code earlier.

Reply:
i think `boost::spirit::qi::parse()` does accept `const_iterator`. and following patch compiles fine in my testbed:
```diff
diff --git a/src/mon/MonCap.cc b/src/mon/MonCap.cc
index c1d7c71230..40ed86be72 100644
--- a/src/mon/MonCap.cc
+++ b/src/mon/MonCap.cc
@@ -601,11 +601,10 @@ struct MonCapParser : qi::grammar<Iterator, MonCap()>

 bool MonCap::parse(const string& str, ostream *err)
 {
-  string s = str;
-  string::iterator iter = s.begin();
-  string::iterator end = s.end();
+  auto iter = str.begin();
+  auto end = str.end();

-  MonCapParser<string::iterator> g;
+  MonCapParser<string::const_iterator> g;
   bool r = qi::parse(iter, end, g, *this);
   //MonCapGrant foo;
   //bool r = qi::phrase_parse(iter, end, g, ascii::space, foo);
```

Reply:
@tchaikov Fixed. I was testing in debian, where it failed. But in Fedora29 it compiles. Probably different versions of `boost::spirit::qi`.

Reply:
we require boost 1.67 and up, and this API is the same in [1.67](https://www.boost.org/doc/libs/1_67_0/libs/spirit/doc/html/spirit/qi/reference/parse_api/iterator_api.html), [1.68](https://www.boost.org/doc/libs/1_68_0/libs/spirit/doc/html/spirit/qi/reference/parse_api/iterator_api.html) and [1.69](https://www.boost.org/doc/libs/1_69_0/libs/spirit/doc/html/spirit/qi/reference/parse_api/iterator_api.html). if we failed to detect the boost version we are using, that'd be a bug.

could you double check it?

Reply:
In debian9.6/stable:
```
jcollin@stratocaster:~$ dpkg -s libboost-dev | grep 'Version'
Version: 1.62.0.1
```
emacs /usr/include/boost/version.hpp:
```
...
#define BOOST_LIB_VERSION "1_62"
```

So it didn't work there.



In Fedora29:

$emacs /home/ceph/build/boost/include/boost/version.hpp
```
...
#define BOOST_LIB_VERSION "1_67"
```

Reply:
Yes, it used the `libboost-dev` installed in the host, ignoring the build/boost. I will file a bug. Thanks.

Reply:
http://tracker.ceph.com/issues/37771